### ISSUE TITLE
Chartmuseum: Projects using names reserved by chartmuseum break harbor

### ISSUE THREAD

**Issue by nlowe** [goharbor/harbor#7068]
## Steps to Reproduce

1. Enable the Chart Museum Feature
2. Search for any project or repository you have access to

### Expected Behavior

* Search results populate as expected

### Actual Behavior

* The API returns a 500 error
  * The following line is logged in `core.log`:

    ```txt
    Mar  4 10:59:33 172.25.0.1 core[1129]: 2019-03-04T15:59:33Z [ERROR] [search.go:141]: failed to filter 
    charts: not found
    ```

* The scroll bar turns red
* The UI shows no results but also does not show any error messages

**Versions:**
Please specify the versions of following systems.
- harbor version: `v1.7.1-c3806525`
- docker engine version: `18.09.3`
- docker-compose version: `1.8.0` 

**Additional context:**

I can provide sanitized config and log files if requested.

This appears to be caused by this check in `src/core/api/search.go`:

https://github.com/goharbor/harbor/blob/c380652540cb8061cff50f0ce2115fc6d217019b/src/core/api/search.go#L139-L143

If the chartmuseum API returns any non-success status code harbor bails with a 500. **This includes `404`'s for charts that don't exist**. Thus, it would appear that in order for search to work, all images in harbor need at least one helm chart added.

I'm also noticing some weird behavior in the chartmuseum logs. It would appear harbor has precreated some projects in chartmuseum as it responds with a `200` for the `index.yaml` for most of the projects on my server, however, some respond with a 404:

```txt
Mar  4 11:14:22 172.25.0.1 chartmuseum[1129]: {"L":"INFO","T":"2019-03-04T16:14:22.799Z","M":"[9417] Request served","path":"/do/index.yaml","comment":"","latency":"25.685884ms","clientIP":"172.27.0.3","method":"GET","statusCode":200,"reqID":"bc3cd5eb-9d2c-46d6-ac3d-cb1c8e6e3b0c"}
Mar  4 11:14:22 172.25.0.1 chartmuseum[1129]: {"L":"WARN","T":"2019-03-04T16:14:22.801Z","M":"[9428] Request served","path":"/api-design/index.yaml","comment":"","latency":"35.42µs","clientIP":"172.27.0.3","method":"GET","statusCode":404,"reqID":"d23eef0a-c55a-41d2-8bfe-aba5e3d2a997"}
```

Is this a botched migration or something? I do not have any helm charts uploaded for either of those projects.

I'm also seeing intermittent pagination issues that seem to be related where some of the pages of projects will respond with a 500 with the following line logged:

```txt
Mar  4 10:54:29 172.25.0.1 core[1129]: 2019-03-04T15:54:29Z [ERROR] [project.go:426]: Failed to get total of charts under project api-design: not found
```

I can make a separate issue for that if required.

**Comment by nlowe** [comment]
This is particularly insideous because of https://github.com/goharbor/harbor/issues/6810#issuecomment-456731876

Apparently chartmuseum is **required** in 1.7 for search to work at all. Looking at the chart server logs I think we're only getting a 404 for that one project so I'm trying to figure out how to manually fix it to restore search for my users.

**Comment by nlowe** [comment]
Actually, we seem to be hitting helm/chartmuseum/issues/201. It's not that the chart doesn't exist (It should return an empty 200 for that), it's that the name `api*` is reserved.

**Comment by nlowe** [comment]
I was able to work around this by stubbing-out chartmuseum:

<details>

<summary>
main.go
</summary>

```golang
package main

import (
    "context"
    "fmt"
    "net/http"
    "os"
    "os/signal"
    "syscall"
    "time"
)

func main() {
    router := http.NewServeMux()
    router.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
        w.Header().Set("Content-Type", "application/x-yaml")
        w.WriteHeader(http.StatusOK)

        w.Write([]byte(fmt.Sprintf(`apiVersion: v1
entries: {}
generated: "%s"
`, time.Now().UTC().Format(time.RFC3339))))
    })

    server := &http.Server{
        Addr: "0.0.0.0:9999",

        WriteTimeout: time.Second * 15,
        ReadTimeout:  time.Second * 15,
        IdleTimeout:  time.Second * 60,

        Handler: router,
    }

    go func() {
        fmt.Println("Starting up on 0.0.0.0:9999")
        if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
            panic(err)
        }
    }()
    
    c := make(chan os.Signal)
	signal.Notify(c, os.Interrupt, syscall.SIGTERM)

	<-c

	fmt.Println("Shutting down...")
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	if err := server.Shutdown(ctx); err != nil {
		fmt.Println("Failed to stop server")
	}
}
```

</details>

By running that in the `docker-compose.chartmuseum.yaml` manifest instead of chartmuseum itself I was able to get search restored. Obviously we won't be able to use chartmuseum now but I'm not as worried about that as I am search not working.

**Comment by nlowe** [comment]
Good news, the latest `master` build of chartmuseum partially fixes this. I built and pushed `nlowe/photon-chartumseum:master` to test. I can now search for projects and browse projects with `api` in their name. However, listing charts for a specific project and uploading charts still do not work, see https://github.com/helm/chartmuseum/issues/201#issuecomment-482176005, I'm not sure if this is a harbor problem or a chartmuseum problem.

**Comment by meyerbro** [comment]
Hey, this is affecting us too, when this is going to be merged to the project? It's not working with goharbor/chartmuseum-photon:v0.8.1-v1.8.0...

**Comment by wphe** [comment]
I confirm I still have the issue with a project that contain `api` in the name. Is it really in the latest version (v1.8.1)?

**Comment by wphe** [comment]
@nlowe can you confirm that this issue is resolved? I still have it (upgrade from v1.7 to v1.8.1).

**Comment by nlowe** [comment]
I haven't had a chance to upgrade my registry to 1.8 yet, but I don't think this made it in the last release.
